decisions/DR009-Batch-method-result-types.md

```diff
@@ -0,0 +1,272 @@
+- **Outcome:** Batch methods will adopt a per-result callback mechanism.
+
+## Background
+
+OpenAssetIO employs a batch-first design. In that, all data centric
+methods take multiple inputs, and produce multiple outputs. This allows
+a `ManagerPlugin` to optimize queries to potentially high-latency
+back-end services.
+
+Many methods may have per-value error states. Such that the output of
+these methods is never of a uniform type. For example, `resolve` for any
+given entity reference may provide a `TraitsData` instance, or an error.
+An error for a specific value should not error the entire batch.
+
+This raises the question of how best express the resulting data across
+the variety of languages supported by OpenAssetIO.
+
+## Relevant data
+
+OpenAssetIO core is currently exposed through C, C++ and Python. The C
+API serves as a binding point for other languages.
+
+A design mantra is that the implementation in any given language should
+be as idiomatic as possible. An exception to which is that Python and
+C++ are often used in the same breath within M&E DCC applications,
+keeping those two languages as close to each other as possible has many
+benefits.
+
+The performance of API methods is important, as many scenarios
+(particularly `resovle`) have a high query volume. Syntax alone is not
+enough of a justification for any particular design choice.
+
+For the scope of this document, it is assumed that the input to any of
+these methods will be an array, regardless of the choices here.
+
+Input options are not explored here, as the sole purpose of the
+batch-centric nature is to optimise queries, which requires all input
+data to be submitted as a single request to back-end services.
+
+## Options considered
+
+### Option 1
+
+Return an array of 'result or error' objects for each input value.
+
+#### Pros
+
+- More familiar in target languages.
+- Facilitates data-centric functional programming without caller-side
+  wrapping.
+- Data-centric result is easier to propagate across process/host
+  boundaries.
+
+#### Cons
+
+- The result type is complex and requires significant boilerplate
+  on the host side to un-pack.
+- Requires additional allocations/copies of the result container.
+
+### Option 2
+
+Provide success/error callbacks to be invoked for each input value.
+
+#### Pros
+
+- More flexible, can be wrapped to look like Option 1 as a
+  host-convenience.
+- Simplifies value handling on the host-side.
+- More easily facilitates future migration to async (by value)
+  operations.
+- Avoids the requirement for all results to be allocated in a single
+  container at any one point in time.
+
+#### Cons
+
+- Less idiomatic in some target languages, especially Python.
+- Precludes data-centric functional programming without caller-side
+  wrapping.
+- Function-centric result is harder to propagate across process/host
+  boundaries.
+- Potential for high per-result call cost depending on the language
+  stack.
+
```

Comment:
```suggestion
- has possible ramifications around concurrency ~ client side code will potentially have to be smart to avoid races and other faults
- at the same time that it has the possibility to simplify value handling, it also has the possibility of just the opposite, causing complexity, under async conditions
```

Reply:
I am definitely buying the argument of performance as outlined in Appendix 2, but it isn't instantly clear to me that the performance characteristics hold up under a a situation such as 200 requests are dispatched to as many threads. I'd be curious to see something of the nature of a worst case

```
condvar<int> resultCount = 0;
mutex serialSorry;
void handler(int requestIndex) { guard(serialSorry); wait(rand()); }
dispatchgroup {
    for (0...199) { dispatchRequest( { wait(rand()); handler(i); ++resultCount; }); }
}
await(resultCount == 200);
```

okay, I guess I have convinced myself of Appendix 2, this worst case example would be no worse than the single threaded example. Move along, nothing to see here :)